Added a new Race field for SayRequest messages, and WebSocket messages.

Cidan opened the pull request:
This PR adds a new field to SayRequest and the associated WebSocket JSON payload for the Race of the speaker. This resolves #200.

Cidan commented:
@karashiiro PTAL. I noticed you previously used unsafe character pointers to get gender for a character. I would also like to remove that code and instead use the safer Customize field, which significantly reduces the complexity of gender lookups and eliminates the need for unsafe.

If that's alright, I'll go ahead and make the change in this PR, along with some more properties I want to add.

karashiiro commented on src/TextToTalk/Backends/SayRequest.cs:
```diff
@@ -29,6 +29,11 @@ public record SayRequest
+    /// <summary>
+    /// The speaker's race in the Race table (e.g. "Hyur", "Elezen", etc.).
+    /// </summary>
+    public required string Race {get; init;}
```
nit: formatting

Cidan replied:
Fixed.

karashiiro commented on src/TextToTalk/TextToTalk.cs:
```diff
@@ -391,6 +402,21 @@ private void BackendSay(SayRequest request)
             this.backendManager.Say(request);
         }
 
+        private string GetSpeakerRace(GameObject? speaker)
```
Overall approach LGTM

karashiiro commented on src/TextToTalk/TextToTalk.cs:
```diff
@@ -391,6 +402,21 @@ private void BackendSay(SayRequest request)
+        private string GetSpeakerRace(GameObject? speaker)
+        {
+            var race = this.data.GetExcelSheet<Race>();
+            if (race is null || speaker is null || speaker is not Character ch)
```
`speaker is null` can be removed from this, the `speaker is not Character ch` handles that already.

karashiiro commented:
LGTM overall, thanks for this!

> I would also like to remove that code and instead use the safer Customize field, which significantly reduces the complexity of gender lookups and eliminates the need for unsafe.

This would be appreciated, I added the original pointer dereferences before Customize had a managed wrapper, which is why it's like that. Now that Dalamud provides a better API around that, these can be removed.

Cidan commented:
Thanks! Unfortunately upon further inspection, it looks like unsafe can't be removed as the model ID is not available via other means I can see.

I'll add some more features here and send this up for an actual review soon.

Cidan commented:
So after talking to a bunch of Dalamud dev folks, it turns out using DrawData is the preferred way of getting this data. This is because other addons might change model information, which wouldn't be reflected in the GameObject, causing mis-matched voices for players with draw data plugins.

I've adjusted the code as such, and added an Age type, based on the model data.

karashiiro commented on src/TextToTalk/TextToTalk.cs:
```diff
@@ -419,6 +426,31 @@ private string GetSpeakerRace(GameObject? speaker)
+            var charaStruct = (FFXIVClientStructs.FFXIV.Client.Game.Character.Character*)speaker.Address;
+            var speakerBodyType = charaStruct->DrawData.CustomizeData.BodyType;
+
+            switch (speakerBodyType)
+            {
+                case 0:
+                    return "Unknown";
+                case 1:
+                    return "Adult";
+                case 3:
+                    return "Elder";
+                case 4:
+                    return "Youth";
+                default:
+                    return "Unknown";
+            }
```
If these aren't coming from a sheet, I'd prefer this being implemented as a `BodyType` enum for the raw data and storing that in `SayRequest`, and then sending this over the wire as a string - Mapperly can do the conversion.

karashiiro commented:
> This is because other addons might change model information, which wouldn't be reflected in the GameObject, causing mis-matched voices for players with draw data plugins.

I was thinking about this while adding the NPC ID field, it might be better to just use that instead of model IDs. Model IDs were a bit of a hack that wound up working when I was first trying to implement the gender overrides. The only issue with that is I'm not sure if enemies use ENpcBase, so it'd only work for regular NPCs.

Anyways, probably out of scope for this, I think.

Cidan commented:
I think for now it probably makes sense to keep doing what you're doing here.

I've added a new enum for body type, renamed the field to body type, and let the JSON mapping take care of converting the name to a string (I had no idea this would work, neat!).

Cidan commented:
I think this is all for now. I have some other ideas, but I want to iterate on what's already implemented.

karashiiro commented on src/TextToTalk/TextToTalk.cs:
```diff
@@ -391,6 +406,38 @@ private void BackendSay(SayRequest request)
+            return row.Masculine;
+        }
+
+        private unsafe BodyType GetSpeakerBodyType(GameObject? speaker)
```
Super non-critical nit: since this doesn't use any instance state, it could be moved to GameObjectExtensions as an extension method (or just to a util file), to declutter the main file a bit.

karashiiro commented:
LGTM, will give it another pass though and merge within the next few hours.
